girder/utility/s3_assetstore_adapter.py

```diff
@@ -172,12 +186,65 @@ def uploadChunk(self, upload, chunk):
             'method': 'PUT',
             'url': url
         }
+
+        return upload
+
+    def _getBucket(self, validate=True):
+        conn = botoConnectS3(self.assetstore['botoConnect'])
+        return conn.lookup(bucket_name=self.assetstore['bucket'],
+                           validate=validate)
+
+    def _proxiedUploadChunk(self, upload, chunk):
+        """
+        Clients that do not support direct-to-S3 upload behavior will go through
+        this method by sending the chunk as a multipart-encoded file parameter
+        as they would with other assetstore types. Girder will send the data
+        to S3 on behalf of the client.
+        """
+        bucket = self._getBucket()
+
+        if upload['s3']['chunked']:
+            if 'uploadId' in upload['s3']:
+                mp = boto.s3.multipart.MultiPartUpload(bucket)
+                mp.id = upload['s3']['uploadId']
+                mp.key_name = upload['s3']['keyName']
+            else:
+                mp = bucket.initiate_multipart_upload(
+                    upload['s3']['key'],
+                    headers=self._getRequestHeaders(upload))
+                upload['s3']['uploadId'] = mp.id
+                upload['s3']['keyName'] = mp.key_name
+                upload['s3']['partNumber'] = 0
+
+            upload['s3']['partNumber'] += 1
+
+            key = mp.upload_part_from_file(
+                chunk, upload['s3']['partNumber'],
+                headers=self._getRequestHeaders(upload))
+            upload['received'] += key.size
+        else:
+            key = bucket.new_key(upload['s3']['key'])
+            key.set_contents_from_file(chunk,
+                                       headers=self._getRequestHeaders(upload))
+
+            if key.size < upload['size']:
```

Comment:
The nature of the proxied uploads is such that we decide at init time whether this is a chunked upload or a single-request upload to S3 based on the CHUNK_LEN property of the adapter. If the upload is less than that, we automatically assume it will be sent in one chunk to girder. If it is not, we let the client know that they should send it all in one chunk.

Given that our chunked upload threshold is 32MB for S3 and our default/recommended chunk size for clients is 64MB, this condition will probably not be reached by regular clients.